### ISSUE TITLE
Timeout and cancellation support

### ISSUE THREAD

**Issue by aksh1618** [dexecutor/dexecutor-core#25]
Hi, thanks for making this amazing library! I'm trying to make a requests execution flow as efficient as possible.

To give an example, say E depends on B and B, C & D depend on A.
```
         A
      /  |   \
    B    C    D
    |
    E
```
Currently, we are using a concept of execution rounds, where we define sets of requests to be executed in parallel, using an executor service and futures, so for this case:

Step 1. Execute A
Step 2. Execute B, C, D
Step 3. Execute E

But we want to do a dependency graph based execution instead, as we want E to be executed as soon as B completes, instead of waiting for all of B, C & D to complete.

For this we are adapting our execution logic to use Dexecutor. But one one missing feature which is stopping us from migrating is the ability to set timeout for and cancel the underlying scheduled task. Is this support planned? Has this been skipped on purpose due to implementation details?

**Comment by mnadeem** [comment]
How about the following ?


````
private static class SleepyTaskProvider implements TaskProvider<String, String> {

    public Task<String, String> provideTask(final String id) {

        return new Task<String, String>() {

            @Override
            public String execute() {
                try {
                    //Perform some task
                    Process process = new ProcessBuilder(command, and, arguments)
    .redirectErrorStream(true)
    .directory(workingDir)
    .start();

process.waitFor(5, TimeUnit.SECONDS);
process.destroy();
process.waitFor(); // wait for the process to terminate
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                String result = id + "processed";
                return result;
            }

            @Override
            public boolean shouldExecute(ExecutionResults<String, String> parentResults) {
                ExecutionResult<String, String> firstParentResult = parentResults.getFirst();
                //Do some logic with parent result
                if ("B".equals(id) && firstParentResult.isSkipped() && firstParentResult.getResult() == "No") {
                    return false;
                }
                return true;
            }
        };          
    }
````

**Comment by aksh1618** [comment]
@mnadeem The tasks in question aren't really processes, they are upstream requests (Being executed using Spring Framework's RestTemplate). Currently we submit it as a Callable to the ExecutorService, and use the returned Future for timeout and cancellation.

**Comment by mnadeem** [comment]
Appreciate if you could provide an example how you are currently doing, and how you would potentially do using dexecutor ( without cancellation support)

**Comment by aksh1618** [comment]
Taking the same example as the original post:
```
    A
 /  |  \
B   C   D
|
E
```
Currently:

```java
var tasks = List.of(
    List.of("A"),
    List.of("B", "C", "D"),
    List.of("E"),
);
tasks.forEach(reqs -> {
    var futures = reqs.stream().map(this::getTaskCallable).map(executorService::submit).collect(toList());
    futures.forEach(future -> {
        future.get(timeout); // Timeout, blocks for max(T_B, T_C, T_D), where T_X is time taken by X
    })
})
```

With Dexecutor:

```java
DexecutorConfig<String, String> config = new DexecutorConfig<>(executorService, new SleepyTaskProvider());
DefaultDexecutor<String, String> executor = new DefaultDexecutor<String, String>(config);

executor.addDependency("A", "B");
executor.addDependency("A", "C");
executor.addDependency("A", "D");
executor.addDependency("B", "E");
executor.execute(ExecutionConfig.NON_TERMINATING);

private static class SleepyTaskProvider implements TaskProvider<String, String> {

    Duration timeout; // set using constructor

    public Task<String, String> provideTask(final String id) {

        return new Task<String, String>() {

            @Override
            public String execute() {
                //Perform some task
                String result = id + "processed";
                return result;
            }

            @Override
            public Duration getTimeout() { // Used internally for interrupting the task with Timeout Exception
                return timeout;
            }

            @Override
            public boolean shouldCancelFutureOnError() { // Used internally for calling cancel on the future in case of Exception
                return true;
            }

            @Override
            public boolean shouldExecute(ExecutionResults<String, String> parentResults) {
                return true;
            }
        };          
    }

```

**Comment by mnadeem** [comment]
Started adding support for timeout 

https://github.com/dexecutor/dexecutor-core/commit/e197f3692fe59eb6d2276dcd4e18c2cd92bfd5db

**Comment by mnadeem** [comment]
Released the following

````
<dependency>
  <groupId>com.github.dexecutor</groupId>
  <artifactId>dexecutor-core</artifactId>
  <version>2.1.3</version>
</dependency>
````

Refer [this example](https://github.com/dexecutor/dexecutor-core/wiki/How-Do-I-%3F#timeout-support)

**Comment by mnadeem** [comment]
Note

Future.cancel(...) delivers an interrupt signal to the thread asking it to stop. You must ensure that your tasks respect the interrupt signals e.g. checks for Thread.currentThread().isInterrupted() at regular intervals.

**Comment by aksh1618** [comment]
Looks great! Really appreciate the quick support! Thanks for making this great tool :heart:

**Comment by mnadeem** [comment]
How about a star for this project, if it has helped you in any way.
